### PR TITLE
Latest draft

### PR DESCRIPTION
**Pull request by Gozala**

### CODE REVIEW

File: src/study-setup.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+// - Replace survey URLs
+// - Figure out how to target English only, United States and Canada
+// - Exclude users with containers extensions to avoid awesome bar conflicts
+// - Ping telemetry on start & on end
+// - Ping on every page load that is an article
+// - Need to handle end if occured through
+//    - navigate away (beforeunload in contentscript ?)
+//    - tab closed (beforeunload in content script ?)
+//    - close window
+//    - quit firefox
```
Comment (ianb) [inline src/study-setup.js]: Are these all still to-do?

I'm assuming that the only surveying this should do is through the popup, and there won't be an end-of-study survey.

File: src/main.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      port.disconnect()
+    }
+  }
+  onReceive(message, port) {
+    if (message.type === "launch-survey") {
+      this.displaySurvey()
+    } else {
+      const payload = {
+        message: "narrate-playback",
+        tab: `${port.sender.tab.id}`,
```
[the remark this answers is not in the document]
Comment (Gozala) [reply src/main.js]: To enable combining playback messages for the same tab, although I'm not certain if we care.

File: src/main.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      port.onDisconnect.addListener(this.onDisconnect)
+      const state = await this.state
+      port.postMessage({ type: "init", state })
+    } else {
+      port.disconnect()
+    }
+  }
+  displaySurvey() {
+    // don't await since we don't really care.
+    browser.storage.sync.set(this.state)
```
[the remark this answers is not in the document]
Comment (Gozala) [reply src/main.js]: It was a good catch! And in fact it was throwing. I believe I used to set state first and then storing it, but for some reason I reversed it before pushing and end up with this.

[the remark this answers is not in the document]
Comment (Gozala) [reply src/main.js]: Fixed by 48d6b88d629c5346e23d2d2a83316921208bcb65

**Pull request merged**